### PR TITLE
Verify image based gadget

### PR DESCRIPTION
**Pull request by eiffel-fl**
Hi.

This PR adds code to verify previously signed image based gadgets:

```bash
$ sudo -E ./ig run ghcr.io/eiffel-fl/gadget/trace_open
INFO[0000] Experimental features enabled                
ERRO[0001] instantiating operator "oci": ensuring image: verifying image "ghcr.io/eiffel-fl/gadget/trace_open": verifying signature: invalid signature when validating ASN.1 encoded signature 
ERRO[0003] instantiating operator "oci": ensuring image: verifying image "ghcr.io/eiffel-fl/gadget/trace_open": verifying signature: invalid signature when validating ASN.1 encoded signature 
^C%
$ sudo -E ./ig run --public-key="$(cat eiffel-fl.pub)" ghcr.io/eiffel-fl/gadget/trace_open
INFO[0000] Experimental features enabled                
RUNTIME.CONTAINERNAME  PID          UID          GID          MNTNS_ID RET FL… MODE        COMM        FNAME                  TIMESTAMP
```

Best regards.

### CODE REVIEW

**Review by blanquicet — COMMENTED** [review]
The logic seems clear to me, we get the cosign image signature for our index, verify it's the signature for our index (DockerManifestDigest) and extract the signature. And finally, we verify that signature with our public key.

Some comments and many questions 😅

File: pkg/oci/oci.go [hunk 1]
```diff
@@ -478,16 +490,141 @@ func newAuthClient(repository string, authOptions *AuthOptions) (*oras_auth.Clie
...
+	if err != nil {
+		return fmt.Errorf("parsing public key: %w", err)
+	}
+
+	verifier, err := signature.LoadVerifier(pub, crypto.SHA256)
+	if err != nil {
+		return fmt.Errorf("loading verifier: %w", err)
+	}
+
+	err = verifier.VerifySignature(bytes.NewReader(decodedSignature), strings.NewReader(payload))
```
Comment (blanquicet) [inline pkg/oci/oci.go:608]: I don't understand why we need to pass the payload. The only useful info I see there is the `docker-manifest-digest` and we already verified it. Is it verified again?

```json
{
  "critical": {
    "identity": {
      "docker-reference": "172.28.128.1:5000/snapshot_socket"
    },
    "image": {
      "docker-manifest-digest": "sha256:4e0a8becac464f4cd37cc871a654639fba6b36c81c6841cede72140231447120"
    },
    "type": "cosign container image signature"
  },
  "optional": null
}
```

File: pkg/oci/oci.go [hunk 2]
```diff
@@ -478,16 +490,141 @@ func newAuthClient(repository string, authOptions *AuthOptions) (*oras_auth.Clie
 	}, nil
 }
 
+func craftSignatureTag(digest string) (string, error) {
+	// WARNING:
+	// https://sigstore.slack.com/archives/C0440BFT43H/p1712253122721879?thread_ts=1712238666.552719&cid=C0440BFT43H
+	// https://github.com/sigstore/cosign/pull/3622
```
[the remark this answers is not in the document]
Comment (blanquicet) [reply pkg/oci/oci.go:496]: Thanks for the explanation. What do you think about adding something like `WARNING: Cosign is considering changing the scheme for publishing/retrieving Sigstore bundles to/from an OCI registry` to the code and keeping the links?

File: pkg/oci/oci.go (hunk 1 above)
Comment (blanquicet) [reply pkg/oci/oci.go:608]: Aahh with that payload we ask the verifier to verify with its pub key and the signature we are passing, the manifest with digest `docker-manifest-digest`. Makes sense. Thanks!

File: pkg/oci/oci.go [hunk 3]
```diff
@@ -478,16 +490,141 @@ func newAuthClient(repository string, authOptions *AuthOptions) (*oras_auth.Clie
...
+	}
+
+	if img.Critical.Image.DockerManifestDigest != imageDigest{
+		return "", "", fmt.Errorf("payload digest does not correspond to image: expected %s, got %s", imageDigest, img.Critical.Image.DockerManifestDigest)
+	}
+
+	return signature, string(payloadBytes), nil
+}
+
+func verifyImage(ctx context.Context, image string, authOpts *AuthOptions) error {
```
[the remark this answers is not in the document]
Comment (blanquicet) [reply pkg/oci/oci.go:582]: Yeah, based on the way we are signing, it seems enough just to verify the index.

File: pkg/oci/oci.go [hunk 4]
```diff
@@ -478,16 +490,141 @@ func newAuthClient(repository string, authOptions *AuthOptions) (*oras_auth.Clie
...
+
+	img := &payload.SimpleContainerImage{}
+	err = json.Unmarshal(payloadBytes, img)
+	if err != nil {
+		return "", "", fmt.Errorf("decoding payload: %w", err)
+	}
+
+	if img.Critical.Image.DockerManifestDigest != imageDigest{
+		return "", "", fmt.Errorf("payload digest does not correspond to image: expected %s, got %s", imageDigest, img.Critical.Image.DockerManifestDigest)
+	}
```
Comment (mauriciovasquezbernal) [inline pkg/oci/oci.go:577]: According to my understanding from https://github.com/containers/image/blob/main/docs/containers-signature.5.md#the-cryptographic-signature, this logic should only be executed after `VerifySignature` returns nil:

> Consumers of container signatures SHOULD verify the cryptographic signature against one or more trusted public keys (e.g. defined in a [policy.json signature verification policy file](https://github.com/containers/image/blob/main/docs/containers-policy.json.5.md)) before parsing or processing the JSON payload in any way, in particular they SHOULD stop processing the container signature if the cryptographic signature verification fails, without even starting to process the JSON payload.

**Review by mauriciovasquezbernal — COMMENTED** [review]
Few quick comments before turning off my PC for the weekend. Will check it out with more care next week.

**Review by blanquicet — COMMENTED** [review]
I still need to test it but looks great from code inspection.

**Review by mauriciovasquezbernal — COMMENTED** [review]
I tested and it looks good.

This work creates the foundations for the image verification, but I think we still need to improve how the user can define the public keys. Something we can do once we have #2493.

**Pull request merged**